### PR TITLE
boring-sys: include all files needed to build FIPS 

### LINKED ISSUE #157: `boring-sys` build fails with CMake unable to find sources when the "fips" feature is enabled
Hi, I'm attempting to build a project that depends on `boring` v3.0.4 in a Docker container, and encountering a build error when the FIPS feature is enabled. 

In both cases, the Dockerfile should contain the requisite build dependencies (note that some of these build deps are not directly relevant to building BoringSSL, but I figured it was helpful to include the whole thing).

<details>

<summary>Dockerfile installed packages</summary>

```dockerfile
RUN DEBIAN_FRONTEND=noninteractive apt-get update
RUN DEBIAN_FRONTEND=noninteractive apt-get install -y curl unzip xz-utils
RUN DEBIAN_FRONTEND=noninteractive apt-get install -y gnupg2
RUN DEBIAN_FRONTEND=noninteractive TZ=Etc/UTC apt-get -y install tzdata
RUN DEBIAN_FRONTEND=noninteractive apt-get install -y git
RUN DEBIAN_FRONTEND=noninteractive apt-get install -y \
  cmake \
  curl \
  gcc \
  software-properties-common \
  gnupg \
  curl \
  apt-transport-https \
  golang-go \
  ca-certificates \
  clang-12 \
  clang-tools-12 \
  clang-format-12 \
  python3-clang-12 \
  libfuzzer-12-dev \
  lldb-12 \
  lld-12 \
  libc++-12-dev \
  libc++abi-12-dev \
  libomp-12-dev
```

</details>

When the "fips" feature flag is not enabled, `boring` (and the rest of the project build successfully. However, when "fips" _is_ enabled, the build script fails with a bunch of CMake errors, complaining about missing sources. Here's the complete output, although it's rather long:

<details>

<summary>Build failure output</summary>

```
#30 19.69 error: failed to run custom build command for `boring-sys v3.0.4`
#30 19.69
#30 19.69 Caused by:
#30 19.70   process didn't exit successfully: `/target/release/build/boring-sys-ae2e1a79ca25e9cc/build-script-build` (exit status: 101)
#30 19.70   --- stdout
#30 19.70   cargo:rerun-if-env-changed=BORING_BSSL_PATH
#30 19.70   cargo:rerun-if-env-changed=BORING_BSSL_INCLUDE_PATH
#30 19.70   cargo:rerun-if-env-changed=BORING_BSSL_SOURCE_PATH
#30 19.70   cargo:rerun-if-env-changed=BORING_SSL_PRECOMPILED_BCM_O
#30 19.70   cargo:rerun-if-env-changed=BORINGSSL_BUILD_DIR
#30 19.70   Initialized empty Git repository in /target/x86_64-unknown-linux-gnu/release/build/boring-sys-36319e5c2e6aa669/out/boringssl-fips/.git/
#30 19.70
#30 19.70   CMAKE_TOOLCHAIN_FILE_x86_64-unknown-linux-gnu = None
#30 19.70   CMAKE_TOOLCHAIN_FILE_x86_64_unknown_linux_gnu = None
#30 19.70   HOST_CMAKE_TOOLCHAIN_FILE = None
#30 19.70   CMAKE_TOOLCHAIN_FILE = None
#30 19.70   CMAKE_GENERATOR_x86_64-unknown-linux-gnu = None
#30 19.70   CMAKE_GENERATOR_x86_64_unknown_linux_gnu = None
#30 19.70   HOST_CMAKE_GENERATOR = None
#30 19.70   CMAKE_GENERATOR = None
#30 19.70   CMAKE_PREFIX_PATH_x86_64-unknown-linux-gnu = None
#30 19.70   CMAKE_PREFIX_PATH_x86_64_unknown_linux_gnu = None
#30 19.70   HOST_CMAKE_PREFIX_PATH = None
#30 19.70   CMAKE_PREFIX_PATH = None
#30 19.70   CMAKE_x86_64-unknown-linux-gnu = None
#30 19.70   CMAKE_x86_64_unknown_linux_gnu = None
#30 19.70   HOST_CMAKE = None
#30 19.70   CMAKE = None
#30 19.70   running: cd "/target/x86_64-unknown-linux-gnu/release/build/boring-sys-36319e5c2e6aa669/out/build" && CMAKE_PREFIX_PATH="" "cmake" "/target/x86_64-unknown-linux-gnu/release/build/boring-sys-36319e5c2e6aa669/out/boringssl-fips" "-DCMAKE_C_COMPILER=clang-12" "-DCMAKE_CXX_COMPILER=clang++-12" "-DCMAKE_ASM_COMPILER=clang-12" "-DFIPS=1" "-DCMAKE_INSTALL_PREFIX=/target/x86_64-unknown-linux-gnu/release/build/boring-sys-36319e5c2e6aa669/out" "-DCMAKE_C_FLAGS= -ffunction-sections -fdata-sections -fPIC -m64" "-DCMAKE_CXX_FLAGS= -ffunction-sections -fdata-sections -fPIC -m64" "-DCMAKE_ASM_FLAGS= -ffunction-sections -fdata-sections -fPIC -m64" "-DCMAKE_BUILD_TYPE=Release"
#30 19.70   -- The C compiler identification is Clang 12.0.0
#30 19.70   -- Check for working C compiler: /usr/bin/clang-12
#30 19.70   -- Check for working C compiler: /usr/bin/clang-12 -- works
#30 19.70   -- Detecting C compiler ABI info
#30 19.70   -- Detecting C compiler ABI info - done
#30 19.70   -- Detecting C compile features
#30 19.70   -- Detecting C compile features - done
#30 19.70   -- The CXX compiler identification is Clang 12.0.0
#30 19.70   -- Check for working CXX compiler: /usr/bin/clang++-12
#30 19.70   -- Check for working CXX compiler: /usr/bin/clang++-12 -- works
#30 19.70   -- Detecting CXX compiler ABI info
#30 19.70   -- Detecting CXX compiler ABI info - done
#30 19.70   -- Detecting CXX compile features
#30 19.70   -- Detecting CXX compile features - done
#30 19.70   -- Found Perl: /usr/bin/perl (found version "5.30.0")
#30 19.70   -- Checking for module 'libunwind-generic'
#30 19.70   --   No package 'libunwind-generic' found
#30 19.70   -- The ASM compiler identification is Clang
#30 19.70   -- Found assembler: /usr/bin/clang-12
#30 19.70   -- Configuring done
#30 19.70
#30 19.70   --- stderr
#30 19.70
#30 19.70   libunwind not found. Disabling unwind tests.
#30 19.70   stat /target/x86_64-unknown-linux-gnu/release/build/boring-sys-36319e5c2e6aa669/out/boringssl-fips/util/godeps.go: no such file or directory
#30 19.70   stat /target/x86_64-unknown-linux-gnu/release/build/boring-sys-36319e5c2e6aa669/out/boringssl-fips/util/godeps.go: no such file or directory
#30 19.70   CMake Error at CMakeLists.txt:565 (add_library):
#30 19.70     Cannot find source file:
#30 19.70
#30 19.70       third_party/googletest/src/gtest-all.cc
#30 19.70
#30 19.70     Tried extensions .c .C .c++ .cc .cpp .cxx .cu .m .M .mm .h .hh .h++ .hm
#30 19.70     .hpp .hxx .in .txx
#30 19.70
#30 19.70
#30 19.70   CMake Error at crypto/test/CMakeLists.txt:23 (add_library):
#30 19.70     Cannot find source file:
#30 19.70
#30 19.70       gtest_main.cc
#30 19.70
#30 19.70     Tried extensions .c .C .c++ .cc .cpp .cxx .cu .m .M .mm .h .hh .h++ .hm
#30 19.70     .hpp .hxx .in .txx
#30 19.70
#30 19.70
#30 19.70   CMake Error at crypto/CMakeLists.txt:489 (add_executable):
#30 19.70     Cannot find source file:
#30 19.70
#30 19.70       abi_self_test.cc
#30 19.70
#30 19.70     Tried extensions .c .C .c++ .cc .cpp .cxx .cu .m .M .mm .h .hh .h++ .hm
#30 19.70     .hpp .hxx .in .txx
#30 19.70
#30 19.70
#30 19.70   CMake Error at crypto/CMakeLists.txt:478 (add_executable):
#30 19.70     Cannot find source file:
#30 19.70
#30 19.70       fipsmodule/rand/urandom_test.cc
#30 19.70
#30 19.70     Tried extensions .c .C .c++ .cc .cpp .cxx .cu .m .M .mm .h .hh .h++ .hm
#30 19.70     .hpp .hxx .in .txx
#30 19.70
#30 19.70
#30 19.70   CMake Error at crypto/test/CMakeLists.txt:1 (add_library):
#30 19.70     Cannot find source file:
#30 19.70
#30 19.70       abi_test.cc
#30 19.70
#30 19.70     Tried extensions .c .C .c++ .cc .cpp .cxx .cu .m .M .mm .h .hh .h++ .hm
#30 19.70     .hpp .hxx .in .txx
#30 19.70
#30 19.70
#30 19.70   CMake Error at ssl/CMakeLists.txt:49 (add_executable):
#30 19.70     Cannot find source file:
#30 19.70
#30 19.70       span_test.cc
#30 19.70
#30 19.70     Tried extensions .c .C .c++ .cc .cpp .cxx .cu .m .M .mm .h .hh .h++ .hm
#30 19.70     .hpp .hxx .in .txx
#30 19.70
#30 19.70
#30 19.70   CMake Error at ssl/CMakeLists.txt:3 (add_library):
#30 19.70     Cannot find source file:
#30 19.70
#30 19.70       bio_ssl.cc
#30 19.70
#30 19.70     Tried extensions .c .C .c++ .cc .cpp .cxx .cu .m .M .mm .h .hh .h++ .hm
#30 19.70     .hpp .hxx .in .txx
#30 19.70
#30 19.70
#30 19.70   CMake Error at ssl/test/CMakeLists.txt:21 (add_executable):
#30 19.70     Cannot find source file:
#30 19.70
#30 19.70       async_bio.cc
#30 19.70
#30 19.70     Tried extensions .c .C .c++ .cc .cpp .cxx .cu .m .M .mm .h .hh .h++ .hm
#30 19.70     .hpp .hxx .in .txx
#30 19.70
#30 19.70
#30 19.70   CMake Error at tool/CMakeLists.txt:3 (add_executable):
#30 19.70     Cannot find source file:
#30 19.70
#30 19.70       args.cc
#30 19.70
#30 19.70     Tried extensions .c .C .c++ .cc .cpp .cxx .cu .m .M .mm .h .hh .h++ .hm
#30 19.70     .hpp .hxx .in .txx
#30 19.70
#30 19.70
#30 19.70   CMake Error at util/fipstools/cavp/CMakeLists.txt:4 (add_executable):
#30 19.70     Cannot find source file:
#30 19.70
#30 19.70       cavp_main.cc
#30 19.70
#30 19.70     Tried extensions .c .C .c++ .cc .cpp .cxx .cu .m .M .mm .h .hh .h++ .hm
#30 19.70     .hpp .hxx .in .txx
#30 19.70
#30 19.70
#30 19.70   CMake Error at util/fipstools/acvp/modulewrapper/CMakeLists.txt:4 (add_executable):
#30 19.70     Cannot find source file:
#30 19.70
#30 19.70       main.cc
#30 19.70
#30 19.70     Tried extensions .c .C .c++ .cc .cpp .cxx .cu .m .M .mm .h .hh .h++ .hm
#30 19.70     .hpp .hxx .in .txx
#30 19.70
#30 19.70
#30 19.70   CMake Error at decrepit/CMakeLists.txt:29 (add_executable):
#30 19.70     Cannot find source file:
#30 19.70
#30 19.70       blowfish/blowfish_test.cc
#30 19.70
#30 19.70     Tried extensions .c .C .c++ .cc .cpp .cxx .cu .m .M .mm .h .hh .h++ .hm
#30 19.70     .hpp .hxx .in .txx
#30 19.70
#30 19.70
#30 19.70   CMake Error at CMakeLists.txt:565 (add_library):
#30 19.70     No SOURCES given to target: boringssl_gtest
#30 19.70
#30 19.70
#30 19.70   CMake Error at crypto/CMakeLists.txt:489 (add_executable):
#30 19.70     No SOURCES given to target: crypto_test
#30 19.70
#30 19.70
#30 19.70   CMake Error at crypto/CMakeLists.txt:478 (add_executable):
#30 19.70     No SOURCES given to target: urandom_test
#30 19.70
#30 19.70
#30 19.70   CMake Error at crypto/test/CMakeLists.txt:23 (add_library):
#30 19.70     No SOURCES given to target: boringssl_gtest_main
#30 19.70
#30 19.70
#30 19.70   CMake Error at crypto/test/CMakeLists.txt:1 (add_library):
#30 19.70     No SOURCES given to target: test_support_lib
#30 19.70
#30 19.70
#30 19.70   CMake Error at ssl/CMakeLists.txt:49 (add_executable):
#30 19.70     No SOURCES given to target: ssl_test
#30 19.70
#30 19.70
#30 19.70   CMake Error at ssl/CMakeLists.txt:3 (add_library):
#30 19.70     No SOURCES given to target: ssl
#30 19.70
#30 19.70
#30 19.70   CMake Error at ssl/test/CMakeLists.txt:21 (add_executable):
#30 19.70     No SOURCES given to target: handshaker
#30 19.70
#30 19.70
#30 19.70   CMake Error at ssl/test/CMakeLists.txt:3 (add_executable):
#30 19.70     No SOURCES given to target: bssl_shim
#30 19.70
#30 19.70
#30 19.70   CMake Error at tool/CMakeLists.txt:3 (add_executable):
#30 19.70     No SOURCES given to target: bssl
#30 19.70
#30 19.70
#30 19.70   CMake Error at util/fipstools/cavp/CMakeLists.txt:4 (add_executable):
#30 19.70     No SOURCES given to target: cavp
#30 19.70
#30 19.70
#30 19.70   CMake Error at util/fipstools/acvp/modulewrapper/CMakeLists.txt:4 (add_executable):
#30 19.70     No SOURCES given to target: modulewrapper
#30 19.70
#30 19.70
#30 19.70   CMake Error at decrepit/CMakeLists.txt:29 (add_executable):
#30 19.70     No SOURCES given to target: decrepit_test
#30 19.70
#30 19.70
#30 19.70   CMake Generate step failed.  Build files cannot be regenerated correctly.
#30 19.70   thread 'main' panicked at '
#30 19.70   command did not execute successfully, got: exit status: 1
#30 19.70
#30 19.70   build script failed, must exit now', /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/cmake-0.1.50/src/lib.rs:1098:5
#30 19.70   note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
#30 19.70 warning: build failed, waiting for other jobs to finish...
```

</details>

I did check the contents of the `google/boring` repo on the FIPS revision (google/boring@853ca1ea1168dff08011e5d42d94609cc0ca2e27), and all the files that CMake complains are missing *appear* to be in the right place. So, I'm not entirely sure what's gone wrong here, or whether it's an issue with the `boring-sys` buildscript or a problem of my own creation. Any help would be very appreciated. Thanks in advance!

### PR DESCRIPTION
**Pull request by hawkw**
This commit modifies the Cargo `include` field for `boring-sys` to
include all the files necessary to actually build the FIPS-certified
revision of BoringSSL. Currently, some of these files are missing (see
#157 for details on this).

This branch improves on my previous approach in PR #158, which switched
from using a Cargo `include` to a Cargo `exclude`. Using `exclude`
rather than `include` resulted in a much larger crates.io package, but
at the time, I thought this was less likely to result in breakage in the
future, because I was concerned about the inability to verify that the
set of excludes/includes can build a new pinned `boringssl` git revision
without having to actually publish a crates.io release.

However, as @nox pointed out in [this comment][1], `cargo package` can
be used to verify a build with the `exclude`s/`include`s applied. This
branch therefore adds `cargo package` steps to CI that check that the
package can actually be built. This way, we are able to make a much
smaller change to the included files, resulting in a smaller package
published to crates.io.

On this branch, the package is 6.7MiB compressed, which is not much
larger than it was previously:

```
    Finished dev [unoptimized + debuginfo] target(s) in 55.65s
    Packaged 1851 files, 33.7MiB (6.7MiB compressed)
```

Fixes #157
Closes #158

[1]: https://github.com/cloudflare/boring/pull/158#issuecomment-1693067112,

### CODE REVIEW

**Comment by inikulin** [comment]
@hawkw  Happy to accept the PR, but there's a catch - `cargo publish` has size limit for package archives and with full Boring crate we hit this limit. With current bundle we're half way through, so we need a test in CI that does `cargo publish` with `--dry-run` to ensure that we stay below the limit. Could you add the test as a part of the PR?

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -197,6 +199,9 @@ jobs:
       run: ln -s clang clang++-12
     - run: cargo test --features fips
       name: Run tests
+    - name: Test boring-sys cargo package (FIPS)
+      run: cargo package -p boring-sys--features fips
```
Comment (hawkw) [inline .github/workflows/ci.yml:203]: whoops:
```suggestion
      run: cargo package -p boring-sys --features fips
```

**Comment by hawkw** [comment]
The Clippy failure (https://github.com/cloudflare/boring/actions/runs/5980651956/job/16306616264#step:10:1) looks unrelated to this change; I'm guessing it's due to new lints added in the latest Rust version. I can open a separate PR to fix that.

**Comment by hawkw** [comment]
@inikulin commit 94e448f6318ae81e1f6c107b3d29a26a84f7dbfc changes the new CI checks to run `cargo publish --dry-run` instead of `cargo package`, since `cargo publish` performs the same verification as `cargo package` in addition to checking the package size, as you requested.

Separately, I've opened #161 to fix the new Clippy lints that are breaking CI. That should hopefully get us a clean build on this branch as well.

**Comment by inikulin** [comment]
@hawkw I think with lints fixes being on master it's safe to merge this one without a rebase on your behalf. Thank you for fix. I'll close #161 now.

**Pull request merged**